launcher/main/main.go

```diff
@@ -60,18 +93,64 @@ func main() {
 		panic(err)
 	}
 
+	for name, secondaryStatic := range staticConfig.Secondaries {
+		if err := launchlib.MkDirs(secondaryStatic.Dirs, stdout); err != nil {
+			fmt.Println("Failed to create directories for secondary process ", name, err)
+			panic(err)
+		}
+	}
+
 	// Compile command
-	cmd, err := launchlib.CompileCmdFromConfig(&staticConfig, &customConfig, stdout)
+	cmds, err := launchlib.CompileCmdsFromConfig(&staticConfig, &customConfig, stdout)
 	if err != nil {
-		fmt.Println("Failed to assemble executable metadata", cmd, err)
+		fmt.Println("Failed to assemble executable metadata", cmds, err)
 		panic(err)
 	}
 
-	// Execute command
-	execErr := syscall.Exec(cmd.Path, cmd.Args, cmd.Env)
+	if len(cmds.Secondaries) != 0 {
+		// Ensure we are in our own process group, since the monitor kills the group
+		if err := syscall.Setpgid(0, 0); err != nil {
+			fmt.Printf("Unable to create process group for process including secondaries")
+			panic(err)
+		}
+
+		monitor := &launchlib.ProcessMonitor{
+			PrimaryPID:      os.Getpid(),
+			ProcessGroupPID: syscall.Getpgrp(),
+		}
+		monitorCmd := exec.Command(os.Args[0], GenerateMonitorArgs(monitor)...)
+		monitorCmd.Stdout = os.Stdout
+		monitorCmd.Stderr = os.Stderr
+
+		// From this point, if the launcher, or subsequent primary process dies, the process group will be terminated
+		// by the process monitor
+		if err := monitorCmd.Start(); err != nil {
+			fmt.Println("Failed to start process monitor for service process group")
+			panic(err)
+		}
+
+		for name, secondary := range cmds.Secondaries {
+			// Ensure child processes are in the same parent group
+			if secondary.SysProcAttr == nil {
+				secondary.SysProcAttr = &syscall.SysProcAttr{}
+			}
+			secondary.SysProcAttr.Setpgid = false
```

Comment:
this is the crucial line to make sure subprocesses _inherit_ the process group id of the launcher